### PR TITLE
LRUcache

### PR DESCRIPTION
**Pull request by satttto**
# PR Details

## What this PR is about

Implemented a least recently used cache.

## TODOs
- [x] Have you assigned "reviewers" and "assignees"?
- [x] Have you written enough comments?
- [x] Have you tested the behaviour?

## What did you use for verifying your implementation?
- [x] Unit Tests
- [x] Unity
- [ ] HoloLens2

## Related links

- Jira: https://lorem-ipsum-ucd23.atlassian.net/jira/software/projects/CLI/boards/1?selectedIssue=CLI-83
- Unit Tests: https://docs.google.com/document/d/1gRJYmDflyA544weU2bgTGZWRlVze3vq-4MBJlKMdubo/edit#heading=h.fp84mhydn43n

### CODE REVIEW

**Review by Peter-Choah — COMMENTED** [review]
From reading the code,
The Put method seems a little buggy.

I don't see an IF statement to check for existing keys in the cache or the linkedlist.

If the cache has a node with **key A** and **Value A** and I put a new node with **key A** and **Value B**, wouldn't it throw a key already added exception instead of updating the key?

Even if it updated the key's value, wouldn't there be multiple nodes with the same key in the linkedlist?

Do you know how I can test C# scripts? I don't know how to confirm on my side.

File: Lorem Ipsum XR Translator/Assets/Scripts/Util/Cache/LRUCache.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            {
+                // Remove the least recently used item from the cache when it's full.
+                var lastNode = this._lruList.First;
+                this._cache.Remove(lastNode.Value.Key);
+                this._lruList.RemoveFirst();
+            }
+
+            var newItem = new LinkedListNode<Item>(new Item { Key = key, Value = val });
+            this._lruList.AddLast(newItem);
+            this._cache.Add(key, newItem);
```
Comment (aconnolly2) [inline Lorem Ipsum XR Translator/Assets/Scripts/Util/Cache/LRUCache.cs]: This doesn't handle updating already existing keys. I suggest modifying this to the following:

`

            if (Get(key) != null) // Item already exists
            {
                var newItem = new LinkedListNode<Item>(new Item { Key = key, Value = val });
                this._lruList.AddLast(newItem);
                this._cache[key] = newItem;
            }
            else // Insert new item
            {
                var newItem = new LinkedListNode<Item>(new Item { Key = key, Value = val });
                this._lruList.AddLast(newItem);
                this._cache.Add(key, newItem);
            }

**Review by aconnolly2 — CHANGES_REQUESTED** [review]

File: Lorem Ipsum XR Translator/Assets/Scripts/Util/Cache/LRUCache.cs (hunk 1 above)
Comment (satttto) [reply Lorem Ipsum XR Translator/Assets/Scripts/Util/Cache/LRUCache.cs]: Yeah, you're right. I should handle the case where the key already exists. I looked it up and the following code creates a new pair of key-value if it doesn't exist.
```c-sharp
this._cache[key] = newItem;
```
I tested it on this playground.
https://dotnetfiddle.net/pznphR

**Comment by satttto** [comment]
Thanks, Peter. You're right. I fixed it. As for the testing, I used this playground but I think it's better to have unit tests like Advanced Data Structure for an util package like this.
https://dotnetfiddle.net/pznphR

@Peter-Choah 
> From reading the code, The Put method seems a little buggy.
> 
> I don't see an IF statement to check for existing keys in the cache or the linkedlist.
> 
> If the cache has a node with **key A** and **Value A** and I put a new node with **key A** and **Value B**, wouldn't it throw a key already added exception instead of updating the key?
> 
> Even if it updated the key's value, wouldn't there be multiple nodes with the same key in the linkedlist?
> 
> Do you know how I can test C# scripts? I don't know how to confirm on my side.

**Comment by satttto** [comment]
I will figure out something.

File: Lorem Ipsum XR Translator/Assets/Scripts/CubeBehavior.cs [hunk 2]
```diff
@@ -29,6 +33,12 @@ void Start()
     {
         Debug.Log("Started");
 
+        // Cache usage. It returns a falsy value if the key doesn't exist.
+        this._cache = new LRUCache<string, string>(10);
+        this._cache.Put("Key", "Value");
+        Debug.Log("Fake Key: " + this._cache.Get("Fake Key"));
+        Debug.Log("Key: " + this._cache.Get("Key"));
+
```
Comment (satttto) [inline Lorem Ipsum XR Translator/Assets/Scripts/CubeBehavior.cs]: @Peter-Choah 
I also put some tests in CubeBehavior.cs

File: Lorem Ipsum XR Translator/Assets/Scripts/Util/Cache/LRUCache.cs (hunk 1 above)
Comment (aconnolly2) [reply Lorem Ipsum XR Translator/Assets/Scripts/Util/Cache/LRUCache.cs]: This looks good and is working in the shiny new unit tests.

**Review by aconnolly2 — APPROVED** [review]

**Pull request merged**